### PR TITLE
[DNA-635] Implement Percentile() aggregation function

### PR DESCRIPTION
**Pull request by dgeelen-uipath**
~This implementation tries to mimic #9, so please compare & contrast with that PR. One will be merged first, so when we merge them we should look at how to combine the engine specific functionality.~ This is now _based on_ #9.

This implements a new SQLKata function `PercentileApprox()`, which generates `APPROX_PERCENTILE` for snowflake, and `PERCENTILE_CONT` for SQL Server (it doesn't have a native percentile approximation). The transform for SQL Server involves converting the main query to a sub-query due to the way partitioning with `PERCENTILE_CONT` works.

### CODE REVIEW

**Review by PG-Thomas — APPROVED** [review]

**Review by MrHacky — CHANGES_REQUESTED** [review]

**Review by MrHacky — CHANGES_REQUESTED** [review]
when talking through this with daniel, we found some more defects in the implementation here

I think we should really consider to stop using sqlite and switch to T-SQL/MS SQL for our tests to prevent wasting effort in 'transpiling' stuff to sqlite, which we don't actually really even use.

File: QueryBuilder.Tests/AggregateTests.cs [hunk 1]
```diff
@@ -1,12 +1,31 @@
...
+                Regex.Replace(
+                    Regex.Replace(
+                        value,
+                        @"\s+",
+                        " "
+                    ),
+                    @"^\s+|\s+$",
+                    ""
+                ),
+                "((\\() | ?(\\)) ?)",
```
[the remark this answers is not in the document]
Comment (dgeelen-uipath) [reply QueryBuilder.Tests/AggregateTests.cs]: Used literal string more consistently.

File: QueryBuilder.Tests/AggregateTests.cs [hunk 2]
```diff
@@ -440,5 +459,123 @@ public void MinAlias()
...
+            Assert.Equal("SELECT APPROX_PERCENTILE(\"c1\", 0.1) AS \"percentileapprox\" FROM \"table\"", Compilers.CompileFor(EngineCodes.Snowflake, query.Clone()).ToString());
+        }
+
+        [Fact]
+        public void SelectPercentileApproxComplex()
+        {
+            var query = new Query()
+                .SelectPercentileApprox("table.c1", 0.1)
+                .From("table")
+                ;
```
[the remark this answers is not in the document]
Comment (dgeelen-uipath) [reply QueryBuilder.Tests/AggregateTests.cs]: All test should now have an implementation, there is a specific test for group-by (and order-by).

File: QueryBuilder/Compilers/SqlServerCompiler.cs [hunk 3]
```diff
@@ -60,8 +65,160 @@ public SqlServerCompiler()
...
+                return new SqlServerAggregatePercentileApproxColumn(this);
+            }
+
+            public override string Compile(SqlResult ctx)
+            {
+                // percentile_cont(0.9) within group (order by "value") over (partition by ...)
+                var column = new Column { Name = Column }.Compile(ctx);
+                var partition = String.Join(", ", GroupByClauses.Select(clause => clause.Compile(ctx)));
+                partition = partition.Length > 0 ? "PARTITION BY " + partition : "";
+                return $"{Type.ToUpperInvariant()}({Percentile}) WITHIN GROUP(ORDER BY {column}) OVER({partition}) {ctx.Compiler.ColumnAsKeyword}{ctx.Compiler.WrapValue(Alias ?? Type)}";
```
Comment (MrHacky) [inline QueryBuilder/Compilers/SqlServerCompiler.cs:89]: `Type' is apparently different here for different implementations.

order by on this aggregate might not work now

**Review by MrHacky — CHANGES_REQUESTED** [review]
seems like this suffers from the same problem we discussed and fixed on https://github.com/UiPath/querybuilder/pull/9 with the class hierarchy.

File: QueryBuilder/Compilers/SqlServerCompiler.cs (hunk 3 above)
Comment (dgeelen-uipath) [reply QueryBuilder/Compilers/SqlServerCompiler.cs]: Fixed by making it a read-only property on the base class.

**Comment by dgeelen-uipath** [comment]
> seems like this suffers from the same problem we discussed and fixed on #9 with the class hierarchy.

Rebased on top of #9 instead.

**Review by MrHacky — CHANGES_REQUESTED** [review]

File: QueryBuilder.Tests/AggregateTests.cs [hunk 4]
```diff
@@ -527,5 +563,298 @@ public void MinAlias()
...
+
+        [Fact]
+        public void SelectPercentileApproxRegularColumns()
+        {
+            var query = new Query()
+                .Select("c1")
+                .SelectAs(("c2", "Alias"))
+                .SelectPercentileApprox(0.9, "c3")
+                .From("table")
+            ;
```
[the remark this answers is not in the document]
Comment (dgeelen-uipath) [reply QueryBuilder.Tests/AggregateTests.cs:683]: Added `group by c1, c2`

File: QueryBuilder/Compilers/SqlServerCompiler.cs [hunk 5]
```diff
@@ -60,8 +65,161 @@ public SqlServerCompiler()
...
+                                                     Component = approxColumn.Component,
+                                                     Distinct = approxColumn.Distinct,
+                                                     Engine = approxColumn.Engine,
+                                                     Percentile = approxColumn.Percentile,
+                                                     GroupByClauses = clauses.Where(x => x is Column group && group.Component == "group").Cast<Column>(),
+                                                 }
+                                            );
+                                            query
+                                                // This MIN acts as a stand-in for ANY_VALUE
+                                                .SelectMin($"{alias}.value_{valueId}", approxColumn.Alias ?? approxColumn.Type)
```
[the remark this answers is not in the document]
Comment (dgeelen-uipath) [reply QueryBuilder/Compilers/SqlServerCompiler.cs:180]: We can now that we've actually rebased on top of that PR

File: QueryBuilder/Compilers/SqlServerCompiler.cs [hunk 6]
```diff
@@ -60,8 +65,160 @@ public SqlServerCompiler()
...
+                        case "select": return 0;
+                        case "group": return 1;
+                        case "order": return 2;
+                        default: return 999;
+                    }
+                })
+                .ToList();
+
+            query.Clauses = new List<AbstractClause>();
+            const string alias = "percentile_approx";
```
Comment (MrHacky) [inline QueryBuilder/Compilers/SqlServerCompiler.cs:128]: shouldn't this be prefixed with some `__` as its an internal identifier we are introducing? (for the subquery at least)

Comment (MrHacky) [reply QueryBuilder/Compilers/SqlServerCompiler.cs:128]: also: also should be the base? column? `Type()`

**Review by MrHacky — CHANGES_REQUESTED** [review]

**Review by MrHacky — APPROVED** [review]

**Pull request merged**